### PR TITLE
Add a SQLite DB delegate that can report to Firebase performance monitoring

### PR DESCRIPTION
**Pull request by vinaysshenoy**
https://app.clubhouse.io/simpledotorg/story/3011/automatically-report-sql-query-performance-to-firebase

### CODE REVIEW

**Comment by vinaysshenoy** [comment]
Given that this is relatively complex, and is quite important to the performance monitoring work, I'd suggest that you pair with @msasikanth while reviewing this PR, @janhavisinghh.

**Comment by janhavisinghh** [comment]
@vinaysshenoy as I'm on leave from tomorrow and Sasikanth is on leave today. Should I reassign it to him?

**Comment by vinaysshenoy** [comment]
> @vinaysshenoy as I'm on leave from tomorrow and Sasikanth is on leave today. Should I reassign it to him?

Yup, that works too.

File: app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package org.simple.clinic.storage.monitoring
+
+import androidx.annotation.VisibleForTesting
+import androidx.annotation.VisibleForTesting.PRIVATE
+
+class DbMethodLookup @VisibleForTesting(otherwise = PRIVATE) constructor(
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt:8]: How do we create an instance of this class outside of the tests? Do we use a factory method?

File: app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    metadataCsv: String
+) {
+
+  private val allDaoMethods: List<DaoMetadata> = DaoMetadata.parse(metadataCsv)
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt:25]: Just a personal preference, we can remove the explicit type?

File: app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    fun matches(daoName: String, lineNumber: Int): Boolean {
+      return this.daoName == daoName && lineNumber in methodRange
+    }
+
+    companion object {
+      fun parse(csv: String): List<DaoMetadata> {
+        return csv
+            .split('\n')
+            .asSequence()
```
Comment (msasikanth) [inline app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt:47]: Any particular reason we converted it to sequence here?

File: app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt (hunk 1 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt:8]: NVM, resolved in next commit.

**Review by msasikanth — APPROVED** [review]
LGTM. Left one comment, can you take a look at it, please?

File: app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt (hunk 2 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt:25]: Yeah, I will make that change in the final PR that integrates the build tooling.

File: app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt (hunk 3 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/storage/monitoring/DbMethodLookup.kt:47]: Not particularly, it was just an IDE recommendation that I auto-approved 😅 .

**Pull request merged**